synapse/federation/sender/__init__.py

```diff
@@ -337,34 +341,85 @@ async def handle_event(event: EventBase) -> None:
                     logger.debug("Sending %s to %r", event, destinations)
 
                     if destinations:
-                        await self._send_pdu(event, destinations)
-
                         now = self.clock.time_msec()
                         ts = await self.store.get_received_ts(event.event_id)
 
                         synapse.metrics.event_processing_lag_by_event.labels(
                             "federation_sender"
                         ).observe((now - ts) / 1000)
 
-                async def handle_room_events(events: Iterable[EventBase]) -> None:
+                        return event, destinations
+                    return None
+
+                async def handle_room_events(
+                    room_id: str, events: Iterable[EventBase]
+                ) -> Tuple[str, Dict[EventBase, Collection[str]]]:
                     with Measure(self.clock, "handle_room_events"):
-                        for event in events:
-                            await handle_event(event)
+                        # Generate event -> destination pairs for every event in events,
+                        # skip if handle_event returns None.
+                        return room_id, {
+                            event: dests
+                            for event, dests in (
+                                ret
+                                for ret in (
+                                    await handle_event(event) for event in events
+                                )
+                                if ret is not None
+                            )
+                        }
```

Comment:
I'm not sure a dict comprehension this large is actually comprehensible (heh), so i added a comment to it, but i'm still not sure if it's not too complex.

Reply:
(Outdated)